euphoria-flink/src/main/java/cz/seznam/euphoria/flink/batch/SortTranslator.java

```diff
@@ -0,0 +1,98 @@
+import cz.seznam.euphoria.core.client.operator.Sort;
+import cz.seznam.euphoria.flink.FlinkOperator;
+import cz.seznam.euphoria.flink.Utils;
+import cz.seznam.euphoria.flink.functions.PartitionerWrapper;
+import cz.seznam.euphoria.shaded.guava.com.google.common.collect.Iterables;
+import org.apache.flink.api.common.operators.Order;
+import org.apache.flink.api.common.typeinfo.TypeHint;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.tuple.Tuple2;
+
+/**
+ * Translator of {@code Union} operator.
+ */
+class SortTranslator implements BatchOperatorTranslator<Sort> {
+  
+  static boolean wantTranslate(Sort operator) {
+    return (operator.getWindowing() == null
+                || (!(operator.getWindowing() instanceof MergingWindowing)
+                    && !operator.getWindowing().getTrigger().isStateful()));
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public DataSet translate(
+      FlinkOperator<Sort> operator,
+      BatchExecutorContext context) {
+    
+    int inputParallelism = Iterables.getOnlyElement(context.getInputOperators(operator)).getParallelism();
+    DataSet input = Iterables.getOnlyElement(context.getInputStreams(operator));
+
+    Sort origOperator = operator.getOriginalOperator();
+
+    final Windowing windowing =
+        origOperator.getWindowing() == null
+            ? AttachedWindowing.INSTANCE
+            : origOperator.getWindowing();
+
+    // extracts partitionId
+    final UnaryFunction<Object, Integer> udfKey = origOperator.getKeyExtractor();
+    final UnaryFunction<Object, Comparable> udfSort = origOperator.getSortByExtractor();
+
+    // ~ extract key/value + timestamp from input elements and assign windows
+    ExtractEventTime timeAssigner = origOperator.getEventTimeAssigner();
+    
+    DataSet<BatchElement> wAssigned =
+        input.flatMap((i, c) -> {
+          BatchElement wel = (BatchElement) i;
+          // assign timestamp if timeAssigner defined
+          if (timeAssigner != null) {
+            wel.setTimestamp(timeAssigner.extractTimestamp(wel.getElement()));
+          }
+          Iterable<Window> assigned = windowing.assignWindowsToElement(wel);
+          for (Window wid : assigned) {
+            Object el = wel.getElement();
+            c.collect(new BatchElement<>(wid, wel.getTimestamp(), el));
+          }
+        })
+        .returns(BatchElement.class)
+        .name(operator.getName() + "::map-input")
+            .setParallelism(inputParallelism);
+    
+    // ~ repartition and sort partitions
+    DataSet<BatchElement> sorted = wAssigned
+        .partitionCustom(new PartitionerWrapper<>(
+            origOperator.getPartitioning().getPartitioner()),
+            Utils.wrapQueryable(we -> udfKey.apply(we.getElement()), Integer.class))
+        .setParallelism(operator.getParallelism())
+        .sortPartition(Utils.wrapQueryable(
+            (BatchElement we) -> Tuple2.of(we.getWindow(), udfSort.apply(we.getElement())), 
```

Comment:
* i think this will again be incredibly slow. `Window` is an opaque type and flink will have to deserialize on every comparison i guess. or am i missing something?
* further, even in the case of knowing the actual window type and being able to register it with flink, we'd also need to make our `Window` impls POJOs or resort to registering special serializers and comparators for them.